src/main/java/dev/openfeature/sdk/ProviderRepository.java

```diff
@@ -106,11 +106,16 @@ private void prepareAndInitializeProvider(@Nullable String clientName,
               BiConsumer<FeatureProvider, String> afterError,
               boolean waitForInit) {
 
+        if (!isProviderRegistered(newProvider)) {
+            // only run afterSet if new provider is not already attached
+            afterSet.accept(newProvider);
+        }
+
         // provider is set immediately, on this thread
         FeatureProvider oldProvider = clientName != null
-            ? this.providers.put(clientName, newProvider)
-            : this.defaultProvider.getAndSet(newProvider);
-        afterSet.accept(newProvider);
+                ? this.providers.put(clientName, newProvider)
+                : this.defaultProvider.getAndSet(newProvider);
+
```

Comment:
Is there a reason you changed the order of these?

Reply:
This is a tricky logic and I couldn't find any good alternative. If we execute `isProviderRegistered` after the collection update, we always get `true` value, making afterSet.accept() not to execute. 

And there's no other good mechanism to detect if the provider is already registered/operational. So I had to make this change.

Reply:
OK - no tests fail so I assume this has no unintended side-effects.

Reply:
It's tests that helped me to improve the logic. Earlier, I had the order which you proposed. But then with tests, I realised we can't check for registration after updating the collections as registration checks rely on the collection content.